File: phase2/frontend/lib/supabaseClient.ts
/**
 * Browser Supabase client using the ANON (public) key only.
 * The service role key never touches the frontend — that lives on the backend.
 *
 * Returns null when env vars are absent so the app runs without auth configured
 * (search is not gated behind auth in Phase 2).
 */
import { createClient, type SupabaseClient } from "@supabase/supabase-js";

const url = process.env.NEXT_PUBLIC_SUPABASE_URL;
const anonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

export const supabase: SupabaseClient | null =
  url && anonKey ? createClient(url, anonKey) : null;

export const isSupabaseConfigured = Boolean(url && anonKey);

File: phase2/backend/supabase_client.py
"""Server-side Supabase client using the SERVICE ROLE key.

The service role key bypasses Row Level Security and must NEVER be exposed to
the browser — it lives only here, on the backend. If Supabase env vars are not
set (e.g. local dev without a Supabase project), `get_supabase()` returns None
and all callers degrade gracefully to a no-cache, no-persistence mode.
"""

from __future__ import annotations

import logging
import os
from functools import lru_cache
from typing import Optional

logger = logging.getLogger(__name__)

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_SERVICE_ROLE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY")


@lru_cache(maxsize=1)
def get_supabase() -> Optional["Client"]:  # noqa: F821 (Client type only when installed)
    """Return a cached service-role Supabase client, or None if not configured."""
    if not SUPABASE_URL or not SUPABASE_SERVICE_ROLE_KEY:
        logger.warning(
            "Supabase not configured (SUPABASE_URL / SUPABASE_SERVICE_ROLE_KEY missing). "
            "Running without cache or search persistence."
        )
        return None
    try:
        from supabase import create_client

        return create_client(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)
    except Exception as exc:  # pragma: no cover - defensive
        logger.error("Failed to initialize Supabase client: %s", exc)
        return None


def is_configured() -> bool:
    return get_supabase() is not None
